doc/developer/design/20240205_cluster_specific_optimization.md

```diff
@@ -0,0 +1,283 @@
+    customers deploy their DDL statements with a script. For the first iteration
+    of this feature it should be therefore sufficient to ask them to run a
+    modified version of this script (with changed catalog item names) against
+    the temporary cluster while running our experiments.
+-   **Support experimentation on unmanaged clusters.** Extending the syntax for
+    managed clusters should be sufficient for an MVP. Since the clusters will be
+    short-lived it doesn't make a huge difference which syntax the Materialize
+    employees will use to create them.
+
+## Solution Proposal
+
+The suggested high-level workflow for assessing the potential impact of an
+optimizer feature will be:
+
+1.  Use `bin/mzexplore` to identify dataflows and clusters that might be affected
+    by an optimizer change gated behind a feature flag.
+2.  For each such cluster, run the following experiment:
+
+The suggested workflow for running an experiment on a specific customer
+environment will be:
+
+1.  Obtain `mz_system` privileges through Teleport.
+2.  Create a dedicated cluster for the experiment. Use the feature flag to
+    enable the tested feature in the `CREATE CLUSTER` definition.
+3.  Create an `UNBILLED` replica for that cluster.
+4.  Ask the customer to replicate (a subset of) dataflow-backed catalog items
+    defined on the the original cluster to the experiment cluster.
+5.  Monitor and record observed differences between the dataflows running in the
+    original cluster and the (modified) dataflows running in the experiment
+    cluster.
+
+<!--
+What is your preferred solution, and why have you chosen it over the
+alternatives? Start this section with a brief, high-level summary.
+
+This is your opportunity to clearly communicate your chosen design. For any
+design document, the appropriate level of technical details depends both on
+the target reviewers and the nature of the design that is being proposed.
+A good rule of thumb is that you should strive for the minimum level of
+detail that fully communicates the proposal to your reviewers. If you're
+unsure, reach out to your manager for help.
+
+Remember to document any dependencies that may need to break or change as a
+result of this work.
+-->
+
+In order to facilitate this workflow, we propose the the following changes
+(discussed in detail below):
+
+-   Extensions to the `CREATE CLUSTER` syntax.
+-   Extensions to the optimizer API.
+
+### Extensions to the `CREATE CLUSTER` syntax
+
+Extend the `CREATE CLUSTER SYNTAX` for managed cluster plans as follows:
+
+```sql
+CREATE CLUSTER cluster
+  REPLICAS (
+    ...
+  )
+  WITH FEATURES (
+    enable_eager_delta_joins = true
+  )
+```
+
+The newly added `WITH FEATURES` clause will be only visible by `mz_system`. We
+will extend `CreateClusterUnmanagedPlan` and `ClusterVariantUnmanaged` with a
+new `ClusterFeatures` struct that models the feature flag configuration that can
+be attached on a per cluster level.
```

Comment:
This sounds like a good place to start.

As future work, it seems like it'd be interesting to allow users to use the `FEATURES` flag, limited to a restricted set of features that are worth allowing users to control. 

We might also consider introducing `CREATE TEMPORARY CLUSTER ... FEATURES (...)`, and limit certain features for use only in temporary clusters. These clusters would only last for a single SQL session, and therefore wouldn't survive an `envd` restart, and therefore wouldn't prevent us from removing the feature flag in a future version of Materialize.

Reply:
:thumbsup: I added this to a "Future work" section at the end of the doc.